# nf-core/clipseq: Output

## :warning: Please read this documentation on the nf-core website: [https://nf-co.re/clipseq/output](https://nf-co.re/clipseq/output)

> _Documentation of pipeline parameters is generated automatically from the pipeline schema and can no longer be found in markdown files._

## Introduction

This document describes the output produced by the pipeline. The plots are taken from the MultiQC report, which summarises results at the end of the pipeline and also includes CLIP-specific summary metrics.

The directories listed below will be created in the results directory after the pipeline has finished. All paths are relative to the top-level results directory.

## Pipeline overview

The pipeline is built using [Nextflow](https://www.nextflow.io/) and processes data using the steps described in the main [README.md](https://github.com/nf-core/clipseq/README.md):

## CLIP summary metrics

The pipeline provides CLIP-specific summary metrics that are plotted in four sections at the begining of MultiQC report:

### Mapping

This section plots the read counts/percentages mapped to the premapping index, to the genome and that remain unmapped.

### Deduplication

This section plots three measures from the UMI-based PCR deduplication.

  1.  *Reads* shows the number of reads before and after deduplication.
  2.  *Ratios* shows the PCR deduplication ratio.
  3.  *Mean UMIs* shows the mean number of unique UMIs per position.

### Crosslinks

This section plots two measures from crosslink identification.

  1.  *Counts* shows the number of crosslinks and crosslink sites.
  2.  *Ratios* shows the ratio of crosslinks to crosslink sites.

### Peaks

This sections plots three peak-calling metrics (if peak calling has been performed) to enable comparison of different tools and optimisation of specific peak-caller parameters.

  1.  *Crosslinks in peaks* shows the total percentage of crosslinks within peaks.
  2.  *Crosslink sites in peaks* shows the total percentage of crosslink sites within peaks.
  3.  *Peak-crosslink coverage* shows the total percentage of nucleotides within peaks that are covered by a crosslink site.

## MultiQC

[MultiQC](http://multiqc.info) is a visualization tool that generates a single HTML report summarizing all samples in your project. Most of the pipeline QC results are visualised in the report and further statistics are available in the report data directory.

The pipeline has special steps which also allow the software versions to be reported in the MultiQC output for future traceability.

For more information about how to use MultiQC reports, see [https://multiqc.info](https://multiqc.info).

**Output files:**

* `multiqc/`
  * `multiqc_report.html`: a standalone HTML file that can be viewed in your web browser.
  * `multiqc_data/`: directory containing parsed statistics from the different tools used in the pipeline.
  * `multiqc_plots/`: directory containing static images from the report in various formats.

## FastQC

[FastQC](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/) gives general quality metrics about your sequenced reads. It provides information about the quality score distribution across your reads, per base sequence content (%A/T/G/C), adapter contamination and overrepresented sequences.

For further reading and documentation see the [FastQC help pages](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/).

**Output files:**

* `fastqc/`
  * `*_fastqc.html`: FastQC report containing quality metrics for your untrimmed raw fastq files.
* `fastqc/zips/`
  * `*_fastqc.zip`: Zip archive containing the FastQC report, tab-delimited data file and plot images.

> **NB:** The FastQC plots displayed in the MultiQC report shows _untrimmed_ reads. They may contain adapter sequence and potentially regions with low quality.

## Pipeline information

[Nextflow](https://www.nextflow.io/docs/latest/tracing.html) provides excellent functionality for generating various reports relevant to the running and execution of the pipeline. This will allow you to troubleshoot errors with the running of the pipeline, and also provide you with other information such as launch commands, run times and resource usage.

**Output files:**

* `pipeline_info/`
  * Reports generated by Nextflow: `execution_report.html`, `execution_timeline.html`, `execution_trace.txt` and `pipeline_dag.dot`/`pipeline_dag.svg`.
  * Reports generated by the pipeline: `pipeline_report.html`, `pipeline_report.txt` and `software_versions.csv`.
  * Documentation for interpretation of results in HTML format: `results_description.html`.
